src/Nethermind/Nethermind.Evm/TransactionProcessing/TransactionProcessor.cs

```diff
@@ -755,44 +755,38 @@ protected virtual GasConsumed Refund(Transaction tx, BlockHeader header, IReleas
             in TransactionSubstate substate, in long unspentGas, in UInt256 gasPrice, int codeInsertRefunds, long floorGas)
         {
             long spentGas = tx.GasLimit;
-            long operationGas = tx.GasLimit;
             var codeInsertRefund = (GasCostOf.NewAccount - GasCostOf.PerAuthBaseCost) * codeInsertRefunds;
 
             if (!substate.IsError)
             {
                 spentGas -= unspentGas;
-                operationGas -= unspentGas;
-                spentGas = Math.Max(spentGas, floorGas);
-                // As per eip-7623, the spent gas is updated to the maximum of the actual gas used or the floor gas,
-                // now we need to recalculate the unspent gas that should be refunded.
-                var unspentGasRefund = tx.GasLimit - spentGas;
 
                 long totalToRefund = codeInsertRefund;
                 if (!substate.ShouldRevert)
                     totalToRefund += substate.Refund + substate.DestroyList.Count * RefundOf.Destroy(spec.IsEip3529Enabled);
                 long actualRefund = RefundHelper.CalculateClaimableRefund(spentGas, totalToRefund, spec);
 
                 if (Logger.IsTrace)
-                    Logger.Trace("Refunding unused gas of " + unspentGasRefund + " and refund of " + actualRefund);
-                // If noValidation we didn't charge for gas, so do not refund
-                if (!opts.HasFlag(ExecutionOptions.SkipValidation))
-                    WorldState.AddToBalance(tx.SenderAddress!, (ulong)(unspentGasRefund + actualRefund) * gasPrice, spec);
+                    Logger.Trace("Refunding unused gas of " + unspentGas + " and refund of " + actualRefund);
                 spentGas -= actualRefund;
-                operationGas -= actualRefund;
             }
             else if (codeInsertRefund > 0)
             {
                 long refund = RefundHelper.CalculateClaimableRefund(spentGas, codeInsertRefund, spec);
 
                 if (Logger.IsTrace)
                     Logger.Trace("Refunding delegations only: " + refund);
-                // If noValidation we didn't charge for gas, so do not refund
-                if (!opts.HasFlag(ExecutionOptions.SkipValidation))
-                    WorldState.AddToBalance(tx.SenderAddress!, (ulong)refund * gasPrice, spec);
                 spentGas -= refund;
-                operationGas -= refund;
             }
 
+            long operationGas = spentGas;
+            spentGas = Math.Max(spentGas, floorGas);
+            long totalRefund = tx.GasLimit - spentGas;
+
+            if (!opts.HasFlag(ExecutionOptions.SkipValidation))
+                WorldState.AddToBalance(tx.SenderAddress!, (ulong)totalRefund * gasPrice, spec);
+            // If noValidation we didn't charge for gas, so do not refund
```

Comment:
Sorry wasn't clear `totalRefund` is only used under `if`, so should be under `if` scope.

Reply:
ohh my bad, that makes more sense actually